module/zfs/dmu.c

```diff
@@ -932,10 +932,10 @@ dmu_free_long_object_impl(objset_t *os, uint64_t object, boolean_t raw)
 	dmu_tx_mark_netfree(tx);
 	err = dmu_tx_assign(tx, TXG_WAIT);
 	if (err == 0) {
-		err = dmu_object_free(os, object, tx);
-		if (err == 0 && raw)
+		if (raw)
 			VERIFY0(dmu_object_dirty_raw(os, object, tx));
```

Comment:
Rather than make this a fatal `VERIFY0` why don't we cleanly handle an error here since we can.  How about:

```c
        err = dmu_tx_assign(tx, TXG_WAIT);
        if (err == 0) {
                if (raw)
                        err = dmu_object_dirty_raw(os, object, tx);
                if (err == 0)
                        err = dmu_object_free(os, object, tx);

                dmu_tx_commit(tx);
        } else {
                dmu_tx_abort(tx);
        }
```

Reply:
I could, but `dmu_object_dirty_raw()` will always succeed, since it will only fail if `dnode_hold()` fails, and we already know at this point in the code that it will work. Do you still want me to change it?

Reply:
I don't see anything in the code which guarantees that.  We do take a hold and release it a few lines up in `dmu_free_long_range()` but by the time we get to `dmu_object_dirty_raw()` and `dmu_object_free` we're not holding any lock or reference which would prevent it from being freed.  An error here is unlikely, but entirely possible.

Incidentally, the same logic applies to the `dmu_object_free()` call which only takes a hold and does handle the error.  So yes, I think it's important we make that change.